api/v1alpha1/workspace_validation.go

```diff
@@ -35,20 +35,152 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
 	if base == nil {
 		klog.InfoS("Validate creation", "workspace", fmt.Sprintf("%s/%s", w.Namespace, w.Name))
 		errs = errs.Also(
+			w.validateCreate().ViaField("spec"),
 			w.Inference.validateCreate().ViaField("inference"),
+			w.Tuning.validateCreate().ViaField("tuning"),
+			w.Tuning.Input.validateCreate().ViaField("input"),
+			w.Tuning.Output.validateCreate().ViaField("output"),
 			w.Resource.validateCreate(w.Inference).ViaField("resource"),
 		)
 	} else {
 		klog.InfoS("Validate update", "workspace", fmt.Sprintf("%s/%s", w.Namespace, w.Name))
 		old := base.(*Workspace)
 		errs = errs.Also(
-			w.Resource.validateUpdate(&old.Resource).ViaField("resource"),
+			w.validateUpdate(old).ViaField("spec"),
 			w.Inference.validateUpdate(&old.Inference).ViaField("inference"),
+			w.Tuning.validateUpdate(&old.Tuning).ViaField("tuning"),
+			w.Tuning.Input.validateUpdate(old.Tuning.Input).ViaField("input"),
+			w.Tuning.Output.validateUpdate(old.Tuning.Output).ViaField("output"),
+			w.Resource.validateUpdate(&old.Resource).ViaField("resource"),
 		)
 	}
 	return errs
 }
 
+func (w *Workspace) validateCreate() (errs *apis.FieldError) {
+	inferenceSpecified := w.Inference.Preset != nil || w.Inference.Template != nil
+	tuningSpecified := w.Tuning.Input != nil
+	if inferenceSpecified != tuningSpecified {
+		errs = errs.Also(apis.ErrGeneric("Either Inference or Tuning must be specified, but not both", ""))
+	}
+	return errs
+}
+
+func (w *Workspace) validateUpdate(old *Workspace) (errs *apis.FieldError) {
+	// Check inference specified
+	oldInferenceSpecified := old.Inference.Preset != nil || old.Inference.Template != nil
+	inferenceSpecified := w.Inference.Preset != nil || w.Inference.Template != nil
+	// Check tuning specified
+	oldTuningSpecified := old.Tuning.Input != nil
+	tuningSpecified := w.Tuning.Input != nil
+	if (!oldInferenceSpecified && inferenceSpecified) || (oldInferenceSpecified && !inferenceSpecified) {
+		errs = errs.Also(apis.ErrGeneric("Inference field cannot be toggled once set", "inference"))
+	}
+
+	if (!oldTuningSpecified && tuningSpecified) || (oldTuningSpecified && !tuningSpecified) {
+		errs = errs.Also(apis.ErrGeneric("Tuning field cannot be toggled once set", "tuning"))
+	}
+	return errs
+}
+
+func (r *TuningSpec) validateCreate() (errs *apis.FieldError) {
+	if r.Input == nil {
+		errs = errs.Also(apis.ErrMissingField("Input"))
+	}
+	if r.Output == nil {
+		errs = errs.Also(apis.ErrMissingField("Output"))
+	}
+	// Currently require a preset to specified, in future we can consider defining a template
+	if r.Preset == nil {
+		errs = errs.Also(apis.ErrMissingField("Preset"))
+	}
+	// TODO: We have to register training plugins and check if it preset exists in plugins here
```

Comment:
isValidPreset, we have this method, right?

Reply:
Yep. Thing is if someone uses an inference preset for performing training? Something to think about

Reply:
We only support fine tuning the model that kaito supports for inference if user specifies the PresetSpec. When we start to support pod template in Tuning spec, the check can be by passed.